src/mono/dbi/cordb-value.cpp

```diff
@@ -0,0 +1,1170 @@
+            char* class_fullname_str = m_dbgprot_decode_string(pReply->p, &pReply->p, pReply->end);
+            int   assembly_id        = m_dbgprot_decode_id(pReply->p, &pReply->p, pReply->end);
+            int   module_id          = m_dbgprot_decode_id(pReply->p, &pReply->p, pReply->end);
+            int   type_id3           = m_dbgprot_decode_id(pReply->p, &pReply->p, pReply->end);
+            int   type_id2           = m_dbgprot_decode_id(pReply->p, &pReply->p, pReply->end);
+            int   token              = m_dbgprot_decode_int(pReply->p, &pReply->p, pReply->end);
+            m_pClass                 = new CordbClass(conn, token, module_id);
+            m_pClass->InternalAddRef();
+            free(namespace_str);
+            free(class_name_str);
+            free(class_fullname_str);
+        }
+
+        m_pCordbType = new CordbType(m_type, conn, NULL, new CordbType((CorElementType)type_id, conn, m_pClass));
+        m_pCordbType->InternalAddRef();
+        m_pCordbType->QueryInterface(IID_ICorDebugType, (void**)ppType);
+        return S_OK;
+    }
+    m_pCordbType = new CordbType(m_type, conn);
+    m_pCordbType->InternalAddRef();
+    m_pCordbType->QueryInterface(IID_ICorDebugType, (void**)ppType);
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::GetSize64(ULONG64* pSize)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbReferenceValue - GetSize64 - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::GetValue(void* pTo)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbReferenceValue - GetValue - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::SetValue(void* pFrom)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbReferenceValue - SetValue - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::IsNull(BOOL* pbNull)
+{
+    if (m_debuggerId == -1)
+        *pbNull = true;
+    LOG((LF_CORDB, LL_INFO1000000, "CordbReferenceValue - IsNull - IMPLEMENTED\n"));
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::GetValue(CORDB_ADDRESS* pValue)
+{
+    if (m_debuggerId == -1)
+        *pValue = NULL;
+    else
+        *pValue = (CORDB_ADDRESS)&m_debuggerId;
+    LOG((LF_CORDB, LL_INFO1000000, "CordbReferenceValue - GetValue - IMPLEMENTED\n"));
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::SetValue(CORDB_ADDRESS value)
+{
+    LOG((LF_CORDB, LL_INFO100000, "CordbReferenceValue - SetValue CORDB_ADDRESS - NOT IMPLEMENTED\n"));
+    return E_NOTIMPL;
+}
+
+HRESULT STDMETHODCALLTYPE CordbReferenceValue::Dereference(ICorDebugValue** ppValue)
+{
+    if (m_debuggerId == -1)
+        return CORDBG_E_BAD_REFERENCE_VALUE;
+    if (m_type == ELEMENT_TYPE_SZARRAY || m_type == ELEMENT_TYPE_ARRAY)
+    {
+        CordbArrayValue* objectValue = new CordbArrayValue(conn, m_pCordbType, m_debuggerId, m_pClass);
```

Comment:
For allocations, we need to make sure we handle the cases when the allocation fails. As of now, I think it will result in a C++ `std::bad_alloc` exception, and we should never throw exceptions across COM boundaries. 

Regarding exceptions, I vaguely recall in our close source debugger code, they have a practice of making sure all COM APIs are wrapped with a blanket catch together with an assert and returning `E_FAIL`. This has been very useful for diagnosing issues.

Reply:
Fixed.